src/lib/cache_lib/impl.ml

```diff
@@ -95,33 +114,47 @@ module Make (Inputs : Inputs_intf) : Intf.Main.S = struct
               ~cache_name:(Cache.name cache) ) ;
       t
 
-    let create cache data =
-      attach_finalizer (Base {data; cache; consumed= false})
+    let create cache data final_state =
+      attach_finalizer (Base {data; cache; transformed= false; final_state})
 
     let assert_not_consumed t msg =
       let open Error in
       if was_consumed t then raise (of_string msg)
 
+    let assert_not_finalized t msg =
+      let open Error in
+      if was_finalized t then raise (of_string msg)
+
     let peek (type a b) (t : (a, b) t) : a =
       assert_not_consumed t "cannot peek at consumed Cached.t" ;
       value t
 
-    let consume (type a b) (t : (a, b) t) : unit =
-      assert_not_consumed t "cannot consume Cached.t twice" ;
-      mark_consumed t
+    let mark_transformed : type a b. (a, b) t -> unit = function
+      | Base x -> x.transformed <- true
+      | Derivative x -> x.transformed <- true
+      | Pure _ -> failwith "cannot set transformed status for pure Cached.t"
 
     let transform (type a b) (t : (a, b) t) ~(f : a -> 'c) : ('c, b) t =
-      consume t ;
+      assert_not_consumed t "cannot consume Cached.t twice" ;
+      mark_transformed t ;
       attach_finalizer
         (Derivative
            { original= original t
            ; mutant= f (value t)
            ; cache= cache t
-           ; consumed= false })
+           ; transformed= false
+           ; final_state= final_state t })
 
-    let invalidate (type a b) (t : (a, b) t) : a Or_error.t =
-      consume t ;
-      let%map () = Cache.remove (cache t) (original t) in
+    let invalidate (type a b) (t : (a, b) t) : a =
+      assert_not_finalized t "Cached item has already been finalized" ;
+      mark_failed t ;
+      Cache.remove (cache t) (original t) ;
+      value t
+
+    let free (type a b) (t : (a, b) t) : a =
+      assert_not_finalized t "Cached item has already been finalized" ;
+      mark_success t ;
```

Comment:
I'm not sure if I find the names `invalidate` and `free` to be semantically accurate here. `invalidate` originally referred to "cache invalidation", which is just the removal of an item from the cache. Now it is being used to be imply a negative, as in there was a failure. Also, `free` seems to imply the release of a resource, which sort of makes sense, but doesn't imply a positive for the success. Perhaps you could pick some better names that directly imply some sort of negative and positive action?